### DISCUSSION TITLE
Changing vendor/product ID in DFU

### DISCUSSION THREAD

**Discussion by victorallume** [micropython/micropython#15793] [STM32 / Pyboard]
I'm wanting to change the  vendor/product ID in the DFU firmware binary so that I can ensure compatibility of the firmware prior to flashing it (which I believe is the original purpose of this field in the DFU spec). I see I can set this via `BOOTLOADER_DFU_USB_VID` and `BOOTLOADER_DFU_USB_PID` in the Makefile. From what I can tell, this is ignored in the flashing process, but I wanted to ask if anyone is aware of any problems that might arise if I change these? Are these used anywhere?

**Comment by bidrohini** [comment]
The Vendor ID (VID) and Product ID (PID) are used to identify the USB device. When the device enters DFU mode, these IDs help the host (e.g., a computer or other flashing tool) recognize and interact with the device correctly.

**Reply by victorallume** [reply]
Yep, and I see that setting BOOTLOADER_DFU_USB_PID/VID would change what the device in mboot mode identifies itself as over USB. Does dfu-util or pydfu care about the usb identifier? Is there a method to get an identifier that isn't already in use, and will play nicely with existing dfu utilities? (btw we don't need to worry about random people plugging USB into a computer - this is just for our own tooling).

Separately, what if, when the mboot dfu is being built from the binary (via `tools/dfu.py`) I specified a different device (e.g. `dfu.pu --device 0xabcd:0xef01`). This only affects the suffix field of the dfu file; the mboot binary that is flashed will identify as what was defined in BOOTLOADER_DFU_USB_PID/VID to the USB. Do any of the dfu utilities use the suffix PID/VID in deciding whether to accept a particular firmware?

**Comment by Josverl** [comment]
VIDs / PIDs are centrally managed, so you can register for a fee.
or for OSS see : https://pid.codes/howto/

**Reply by victorallume** [reply]
Sure, though as far as I can tell, the VID for a heap of micropython boards is the default 0xFF05 which isn't registered.

My question is more specifically regarding the PID and VID in a DFU binary file. The stated purpose for these fields in the suffix is to ensure the DFU  binary is compatible with the hardware that is applying it, which seems to be a different purpose to identifying over USB, even though it is using the same naming. For example, there are a bunch of different PIDs in the stm32 port for different combinations of VCPs and MSC. And many the stm32 boards use the same VID and PID even though they're on incompatible hardware; additionally, the mboot VID/PID is the same for all stm32 boards.

**Reply by Josverl** [reply]
I think I see the unclarity you point out, 
but I have insufficient experience with building for different stm32 boards to be of much help in this.
Once you have clarity, this seems something worth improving the documentation on, or add a quick page in the wiki.
